[EDITS]Copy to edit (#1244)
mulesoft/docs-connectors: edifact-edi/2.2/modules/ROOT/edits/index.adoc

= EDIFACT EDI Connector 2.2 - Mule 4
:page-aliases: connectors::edifact/edifact-edi-connector.adoc

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Premium]

Anypoint Connector for EDIFACT EDI (EDIFACT Connector) enables you to convert EDIFACT messages to and from DataWeave-compatible representations using lists and maps.

> The  xref:edifact-edi-versions.adoc[supported EDIFACT versions] are provided.

Reviewer: passive

> The  xref:edifact-edi-versions.adoc[supported EDIFACT versions] are provided.

Reviewer: ```suggestion
The xref:edifact-edi-versions.adoc[supported EDIFACT versions] are provided.
```

== Prerequisites

To use this connector, you must be familiar with:

* Anypoint Connectors
* Mule runtime engine (Mule)
* Elements and global elements in a Mule flow
* Creating a Mule app using Anypoint Studio (Studio)

To use EDIFACT EDI Connector, you must have:

> * Purchased a MuleSoft license for Anypoint B2B

Reviewer: ```suggestion
* A MuleSoft license for Anypoint B2B
```

* Anypoint Studio 7.0 or later

> [NOTE]

Reviewer: I probably commented on this being a Note in the previous edit. It seems too important to be a Note at the end of the prereqs and I wonder if there is not a prereq category for its content: could it be incorporated into "you must have" as "Memory sufficient to meet the requirement for..."?

> EDIFACT Connector supports files of up to 15 MB in size, and the memory requirement is approximately 40 to 1.

Reviewer: Implement previous edit wordings for this same text.
Also, beware extra spaces.

> For example, a 1-MB file requires up to 40 MB of memory to process, so it's important to consider this memory requirement in conjunction with your Transactions Per Second (TPS)  needs for large X12 files.

Reviewer: ```suggestion
For example, a 1-MB file requires up to 40 MB of memory to process, so it's important to consider this memory requirement in conjunction with your Transactions Per Second (TPS) needs for large X12 files.
```

This is not an exact figure; the value varies according to the complexity of the mapping instructions.

=== Understanding the EDI Schema Language

> The EDIFACT EDI Connector uses a YAML format called EDI Schema Language (ESL) to represent EDI schemas. Basic ESLs define the structure of EDI messages in terms of

Reviewer: ```suggestion
The EDIFACT EDI Connector uses a YAML format called EDI Schema Language (ESL) to represent EDI schemas. Basic ESLs define the structure of EDI messages in terms of:
```

> * structures, known as messages, in EDIFACT terminology
> 
> * groups, segments, composites, and elements

Reviewer: Begin each with init. cap. Also:
```suggestion
* Structures, known as _messages_ in EDIFACT terminology

* Groups, segments, composites, and elements
```

> You can customize the EDIFACT schema definitions to suit your data by defining an overlay schema. An overlay schema is a type of ESL that enables you to modify a base schema, such as an EDIFACT D.98B ORDERS schema, with your specific conventions. You don't need an overlay schema if you're using the structure defined by the standard, but most EDI exchanges modify the base definitions in an implementation convention.

Reviewer: ```suggestion
You can customize the EDIFACT schema definitions to suit your data by defining an _overlay schema_. An overlay schema is a type of ESL that enables you to modify a base schema, such as an EDIFACT D.98B ORDERS schema, with your specific conventions. You don't need an overlay schema if you're using the structure defined by the standard, but most EDI exchanges modify the base definitions in an implementation convention.
```

Reviewer: Do we need to explain this further:
"but most EDI exchanges modify the base definitions in an implementation convention."

> You can also define your own schemas from scratch, or copy and edit a base EDIFACT schema definition to suit your data as shown in the  xref:x12-edi-connector::x12-edi-schema-language-reference.adoc[EDI Schema Language Reference].

Reviewer: GLOBALLY rewrite to avoid idiomatic language such as "from scratch."

Reviewer: Remove extra space.

== Finding Schema Locations

To use the connector, you need to know the locations of the schemas
in your project. If you're using out-of-the-box EDIFACT schemas and
not customizing anything, the schema location follows the
`/edifact/{version}/{message}.esl` convention.
For example, using the D.98B version of the ORDERS message, your schema location is `/edifact/d98b/ORDERS.esl`.

If you're using one or more custom schemas, place these in the
`src/main/resources` directory and refer to the location relative to this directory.
For example, if you've put your ORDERS schema (either overlay or full) in `src/main/resources/mypartner/ORDERS.esl`,
your schema location is `/mypartner/ORDERS.esl`.

> 

Reviewer: How many types of schemas are there: overlay, basic, full? In addition to out-of-the-box and custom?
Be sure that each term is made distinct; omit one of any pair or group that is used synonymously.
GLOBAL in all writing.

== The Hierarchy of EDIFACT Connector Message Structure

> Use this connector to read or write EDIFACT documents into (or from) the canonical EDI message structure. This structure is a hierarchy of Java Maps and Lists. You use either DataWeave or code to manipulate them. Each transaction has its own structure as defined in the schemas.

Reviewer: ```suggestion
Use this connector to read or write EDIFACT documents to (or from) the canonical EDI message structure. This structure is a hierarchy of Java Map and List objects. You use either DataWeave or code to manipulate them. Each transaction has its own structure, as defined in the schemas.
```

> The message contains the following keys, some of which apply to either the Read operation or Write operation, as indicated:

Reviewer: ```suggestion
The message contains the following keys, some of which apply to either the Read operation or the Write operation, as indicated:
```

Reviewer: See above, and in the following table Description column, consistently capitalize Read operation and Write operation.

[%header%autowidth.spread]
|===
|Key Name |Description

> |Delimiters |An optional set of delimiters used in the message. If used in a read operation, the delimiters are based on the most recent interchange processed by a read operation. If used for a write operation, the delimiters override the values in the module configuration. The characters in the string of delimiters are interpreted based on their position in string, in the following order:
> data element separator, component element separator, repetition separator, segment terminator, release character.

Reviewer: ```suggestion
|Delimiters |An optional set of delimiters used in the message. If used in a read operation, the delimiters are based on the most recent interchange processed by a read operation. If used for a write operation, the delimiters override the values in the module configuration. The characters in the string of delimiters are interpreted based on their position in string, in the following order: data element separator, component element separator, repetition separator, segment terminator, and release character.
```

|Messages |A hierarchy of messages that are either read by the module during a read operation or sent by the module during a write operation. The value of this top-level key is a map with standard versions as keys, in the form `D96A`. For example, if you are using version `D96.A ORDERS` and `CUSRES` messages, the Messages contain a map with one key, `D96A`. The value of this key is another map, one with two keys, `ORDERS` and `CUSRES`. Each of these contains a list of individual ORDERS and CUSRES messages that are processed.
|===

Individual messages have their own maps, with the following keys:

[%header%autowidth.spread]
|===
|Key |Description
|Detail |Map of segments or loops from the detail section of the message. Values are maps for segments or loops that occur once. Lists of maps are for values that occur more than once.
|Errors (read only) |A list of errors associated with the message.
|Heading |A map of segments or loops from the heading section of the message. Values are maps for segments or loops that occur once. Lists of maps exist for values that occur more than once.
|Id |The Message ID must match the key of the containing message list.

> |Interchange |A map of UNB interchange header segment values. For a read message, the map is the actual data from the enclosing interchange. It is a single map linked from all messages in the interchange. For a write message, the map value creates the enclosing interchange. For example, gathering messages with the same interchange values into a single interchange, regardless of whether the actual maps are the same. If there are no values for this map, then the default is the value(s) taken from the Interchange map at the message level.

Reviewer: Replace value(s) with the most likely of value or values; never use this (s) structure.

> |MessageHeader |Map of UNH message header segment values. This provides the actual header data for a Read operation, and enables you to provide overrides for for Write operation configuration settings.

Reviewer: ```suggestion
|MessageHeader |Map of UNH message header segment values. This provides the actual header data for a Read operation, and enables you to provide overrides for Write operation configuration settings.
```

|Name |Message name
|Summary |Map of segments or loops from the summary section of the message. Values are maps for segments or loops that occur once. Lists of maps are for values that occur more than once.
|===

> Generated CONTRL functional acknowledgment messages differ from Received messages in their handling of interchange information:

Reviewer: GLOBALLY standardize between Received messages and received messages.

[%header%autowidth.spread]
|===
|Key name |Description

> |Interchange |For functional acknowledgments generated by Receive processing, this map is a copy of the data for the containing interchange with Sender and Receiver identification components (UNB2.1/UNB2.2 and UNB2.1/UNB3.2). When writing a message, these values are used for creating the enclosing interchange. For example, gathering messages with the same interchange values into a single interchange, regardless of whether the actual maps are the same. If there are no values for this map, then the default is the value(s) taken from the Interchange map at the message level.

Reviewer: unclear reference: "When writing a message, these values are used...." What is writing the message?

Reviewer: Fragment:
For example, gathering messages with the same interchange values into a single interchange, regardless of whether the actual maps are the same.

Reviewer: ditto re. value or values rather than value(s); GLOBAL
